### ISSUE TITLE
conda install -f broken (when redownloading) in 4.0.0

### ISSUE THREAD

**Issue by ilanschnell** [conda/conda#2179]
So @kalefranz, I found the issue!  The problem is because of the new code starting at
https://github.com/conda/conda/blob/master/conda/api.py#L44
When the exact same package is updated on the server, the metadata still comes from the local metadata in the `<PREFIX>/conda-meta/` directory.  This is really bad, because then the download will fail with an `MD5 mismatch error`, because the new package on the server has a new MD5.

@mcg1969, I thought you told me the local data from `<PREFIX>/conda-meta/` wasn't used anymore, unless no official repodata was available.  This is not the case.  We need to fix this ASAP.

**Comment by mcg1969** [comment]
I don't recall saying that. Here is the strategy the new code employing.

--- If the MD5 signatures match between the repodata and the installed package, _prefer the repodata_.
--- If there is a mismatch, prefer the _installed package's_ repodata.
--- However, if the installed package has no `depends` field, go ahead and pull it from the repodata instead.

So the installed metadata takes priority only if the package on the server does not match, _and_ it has non-empty depdendencies.

So my question to you is why is the MD5 signature changing _without the build number being updated_?

Please feel free to fix this however you wish though. If I am hearing you correctly, you're simply going to want to simplify the logic so that the repodata is preferred _regardless_ of MD5. That's fine.

**Comment by ilanschnell** [comment]
The MD5 can change (without changing the build number) if one does a force uploaded of a conda package to anaconda.org.  I'm not sure how often this is the case, but it happens, and if it does the user has currently no ability to get the new package.

I have a question for you as well.  How does the new solver look at the installed packages?  I thought it did that by using `installed.linked_data()`, but this function is only used in `api.get_index()`.

**Comment by mcg1969** [comment]
Great. Then my recommendation is you bypass my MD5 check in the above code.

The solver _only_ looks at the output `api.get_index()`. That's why it is essential that the installed packages be "slipstreamed" into that index dictionary. So lines 44-52 need to be kept, even if we change the rules about  updating.

**Comment by mcg1969** [comment]
Something as simple as this might accomplish what you're hoping:
```
        for fn, info in iteritems(install.linked_data(prefix)):
            fn = fn + '.tar.bz2'
            index.setdefault(fn, info)
```

**Comment by ilanschnell** [comment]
Thanks, @mcg1969.  I am still unclear how this works.  The lines 44-52 only add items to the `index` dict (if any).  How does the solver know whether something was in the initial index (from `fetch_index()`) or whether the item as added in line 52 ?  I just don't see how the solver can possibly know what.

**Comment by mcg1969** [comment]
The solver is given a list of installed dists along with the specs. It assumes that the metadata for those installed dists is in the index.

**Comment by mcg1969** [comment]
In other words, the solver is _depending_ on the data in the index providing accurate guidance for both installed packages and non-installed packages.

**Comment by ilanschnell** [comment]
Ah, OK, thanks.  Now it makes more sense to me.

**Comment by ilanschnell** [comment]
I've created this PR https://github.com/conda/conda/pull/2183 for a possible fix.  Needs testing.

**Comment by mcg1969** [comment]
Great. I think we need to get this out, as I think it's causing the problems with #2181 as well, because the internal metadata for Anaconda 2.5.0 is inconsistent, and cannot be satisfied. In particular:

- `pickleshare-0.5-py35_0.tar.bz2` requires `path.py 7.6.1` internally, but Anaconda 2.5 ships with `path.py 8.1.2`. The repodata.json file fixes this by relaxing the requirement to simply `path.py`
- `python-3.5.1-0.tar.bz2` requires `openssl 1.0.2d` internally, but Anaconda 2.5 ships with `openssl 1.0.2f`. The repodata.json file relaxes this to `openssl 1.0.2*`.

**Comment by kalefranz** [comment]
Should be fixed with #2183.
